lib/Dialect/Arc/Transforms/LowerToLLVM.cpp

```diff
@@ -0,0 +1,471 @@
+  }
+};
+
+struct AllocStateOpLowering : public ConversionPattern {
+  AllocStateOpLowering(TypeConverter &typeConverter, MLIRContext *context,
+                       PatternBenefit benefit = 1)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), benefit,
+                          context) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const final {
+    if (!isa<AllocStateOp, RootInputOp, RootOutputOp>(op))
+      return failure();
+
+    // Get a pointer to the correct offset in the storage.
+    auto offsetAttr = op->getAttr("offset");
+    assert(offsetAttr && "cannot lower unallocated state to LLVM");
+    Value offset = rewriter.create<LLVM::ConstantOp>(
+        op->getLoc(), rewriter.getI32Type(), offsetAttr);
+    Value ptr = rewriter.create<LLVM::GEPOp>(
+        op->getLoc(), operands[0].getType(), operands[0], offset);
+
+    // Cast the raw storage pointer to a pointer of the state's actual type.
+    auto ptrType = LLVM::LLVMPointerType::get(
+        op->getResult(0).getType().cast<StateType>().getType());
+    if (ptrType != ptr.getType())
+      ptr = rewriter.create<LLVM::BitcastOp>(op->getLoc(), ptrType, ptr);
+
+    // Load the value.
+    rewriter.replaceOp(op, ptr);
+    return success();
+  }
+};
+
+struct StateReadOpLowering : public OpConversionPattern<arc::StateReadOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::StateReadOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    rewriter.replaceOpWithNewOp<LLVM::LoadOp>(op, adaptor.getState());
+    return success();
+  }
+};
+
+struct StateWriteOpLowering : public OpConversionPattern<arc::StateWriteOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::StateWriteOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    if (adaptor.getCondition()) {
+      rewriter.replaceOpWithNewOp<scf::IfOp>(
+          op, adaptor.getCondition(), [&](auto &builder, auto loc) {
+            builder.template create<LLVM::StoreOp>(loc, adaptor.getValue(),
+                                                   adaptor.getState());
+            builder.template create<scf::YieldOp>(loc);
+          });
+    } else {
+      rewriter.replaceOpWithNewOp<LLVM::StoreOp>(op, adaptor.getValue(),
+                                                 adaptor.getState());
+    }
+    return success();
+  }
+};
+
+struct AllocMemoryOpLowering : public OpConversionPattern<arc::AllocMemoryOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::AllocMemoryOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    Value offset = rewriter.create<LLVM::ConstantOp>(
+        op.getLoc(), rewriter.getI32Type(), op->getAttr("offset"));
```

Comment:
Yeah we should overhaul this a bit at some point. Especially since there are four ops that want some form of allocatable offset, and it's also not clear if the version with an offset should be a separate op, for more strict guarantees in the IR :thinking: Or if this should be somewhere in the type?

Reply:
Yeah good point. I agree that a more strict separation between allocated and non-allocated operations would be nice.